odoo/addons/base/models/ir_attachment.py

```diff
@@ -50,22 +50,62 @@ def _storage(self):
     @api.model
     def _filestore(self):
         return config.filestore(self._cr.dbname)
-
+    
+    @api.model
+    def _get_datas_inital_vals(self):
+        return {
+            'store_fname': False,
+            'db_datas': False,
+        }
+    
+    @api.model
+    def _update_datas_vals(self, vals, attach, bin_data):
+        vals.update({
+            'file_size': len(bin_data),
+            'checksum': self._compute_checksum(bin_data),
+            'index_content': self._index(bin_data, attach.datas_fname, attach.mimetype),
+        })
+        return vals
+    
+    @api.model
+    def _get_datas_clean_vals(self, attach):
+        vals = {}
+        if attach.store_fname:
+            vals['store_fname'] = attach.store_fname
+        return vals
+    
+    @api.model
+    def _clean_datas_after_write(self, vals):
+        if 'store_fname' in vals:
+            self._file_delete(vals['store_fname'])
+    
+    @api.model
+    def storage_locations(self):
+        return ['db', 'file']
+    
     @api.model
     def force_storage(self):
-        """Force all attachments to be stored in the currently configured storage"""
         if not self.env.user._is_admin():
             raise AccessError(_('Only administrators can execute this action.'))
-
-        # domain to retrieve the attachments to migrate
-        domain = {
-            'db': [('store_fname', '!=', False)],
-            'file': [('db_datas', '!=', False)],
-        }[self._storage()]
-
-        for attach in self.search(domain):
-            attach.write({'datas': attach.datas})
+        storage_domain = {
+            'db': ('db_datas', '=', False),
+            'file': ('store_fname', '=', False), 
+        }
+        record_domain = [
+            '&', ('type', '=', 'binary'),
```

Comment:
that does not seems right, you will get something like
```python
[ '&', ('type', '=', 'binary'), '&', ('store_fname', '=', False), '|', ('res_field', '=', False), ('res_field', '!=', False)]
```

Reply:
Isn't that the same as this form `[ '&', '&', ('type', '=', 'binary'), ('store_fname', '=', False), '|', ('res_field', '=', False), ('res_field', '!=', False)]` only in a different notation